File: LeetCode/Java Solutions/Maximum Score of a Good Subarray/maximum.java
import java.util.*;

class Solution {
    public int maxScoreWords(String[] words, char[] letters, int[] score) {
        Map<Character, Integer> letterCount = new HashMap<>();
        for (char letter : letters) {
            letterCount.put(letter, letterCount.getOrDefault(letter, 0) + 1);
        }
        
        return backtrack(words, score, letterCount, 0, 0);
    }
    
    private int backtrack(String[] words, int[] score, Map<Character, Integer> letterCount, int index, int currentScore) {
        if (index == words.length) {
            return currentScore;
        }
        
        int maxScore = backtrack(words, score, letterCount, index + 1, currentScore);
        
        String word = words[index];
        if (canFormWord(word, letterCount)) {
            int wordScore = getWordScore(word, score);
            for (char ch : word.toCharArray()) {
                letterCount.put(ch, letterCount.get(ch) - 1);
            }
            maxScore = Math.max(maxScore, backtrack(words, score, letterCount, index + 1, currentScore + wordScore));
            for (char ch : word.toCharArray()) {
                letterCount.put(ch, letterCount.get(ch) + 1);
            }
        }
        
        return maxScore;
    }
    
    private boolean canFormWord(String word, Map<Character, Integer> letterCount) {
        Map<Character, Integer> wordCount = new HashMap<>();
        for (char ch : word.toCharArray()) {
            wordCount.put(ch, wordCount.getOrDefault(ch, 0) + 1);
        }
        for (Map.Entry<Character, Integer> entry : wordCount.entrySet()) {
            if (letterCount.getOrDefault(entry.getKey(), 0) < entry.getValue()) {
                return false;
            }
        }
        return true;
    }
    
    private int getWordScore(String word, int[] score) {
        int totalScore = 0;
        for (char ch : word.toCharArray()) {
            totalScore += score[ch - 'a'];
        }
        return totalScore;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.maxScoreWords(new String[]{"dog","cat","dad","good"}, new char[]{'a','a','c','d','d','d','g','o','o'}, new int[]{1,0,9,5,0,0,3,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0}));  // Output: 23
        System.out.println(sol.maxScoreWords(new String[]{"xxxz","ax","bx","cx"}, new char[]{'z','a','b','c','x','x','x'}, new int[]{4,4,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,0,10}));  // Output: 27
        System.out.println(sol.maxScoreWords(new String[]{"leetcode"}, new char[]{'l','e','t','c','o','d'}, new int[]{0,0,1,1,1,0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0}));  // Output: 0
    }
}

File: LeetCode/Python Solutions/Maximum Score Words Formed by Letters/maximum.py
from typing import List
from collections import Counter

class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        def get_word_score(word: str) -> int:
            return sum(score[ord(ch) - ord('a')] for ch in word)
        
        def can_form_word(word: str, letter_count: Counter) -> bool:
            word_count = Counter(word)
            for ch in word_count:
                if word_count[ch] > letter_count[ch]:
                    return False
            return True
        
        def backtrack(index: int, current_score: int, letter_count: Counter) -> int:
            if index == len(words):
                return current_score
            
            max_score = backtrack(index + 1, current_score, letter_count)
            
            word = words[index]
            if can_form_word(word, letter_count):
                word_score = get_word_score(word)
                for ch in word:
                    letter_count[ch] -= 1
                max_score = max(max_score, backtrack(index + 1, current_score + word_score, letter_count))
                for ch in word:
                    letter_count[ch] += 1
            
            return max_score
        
        letter_count = Counter(letters)
        return backtrack(0, 0, letter_count)

# Example usage
sol = Solution()
print(sol.maxScoreWords(["dog","cat","dad","good"], ["a","a","c","d","d","d","g","o","o"], [1,0,9,5,0,0,3,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0]))  # Output: 23
print(sol.maxScoreWords(["xxxz","ax","bx","cx"], ["z","a","b","c","x","x","x"], [4,4,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,0,10]))  # Output: 27
print(sol.maxScoreWords(["leetcode"], ["l","e","t","c","o","d"], [0,0,1,1,1,0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,0,0,0]))  # Output: 0




